Python/optimizer_analysis.c

```diff
@@ -1,50 +1,1081 @@
+            if (sym_is_type(local, NULL_TYPE)) {
+                goto error;
             }
-            case _CHECK_PEP_523:
-            {
-                /* Setting the eval frame function invalidates
-                 * all executors, so no need to check dynamically */
-                if (_PyInterpreterState_GET()->eval_frame == NULL) {
-                    buffer[pc].opcode = _NOP;
+            PEEK(1) = local;
+            break;
+        }
+        case LOAD_FAST: {
+            STACK_GROW(1);
+            _Py_UOpsSymType * local = GETLOCAL(oparg);
+            if (sym_is_type(local, NULL_TYPE)) {
+                Py_UNREACHABLE();
+            }
+            // Guaranteed by the CPython bytecode compiler to not be uninitialized.
+            PEEK(1) = GETLOCAL(oparg);
+            assert(PEEK(1));
+
+            break;
+        }
+        case LOAD_FAST_AND_CLEAR: {
+            STACK_GROW(1);
+            PEEK(1) = GETLOCAL(oparg);
+            break;
+        }
+        case LOAD_CONST: {
+            STACK_GROW(1);
+            PEEK(1) = (_Py_UOpsSymType *)GETITEM(
+                ctx, oparg);
+            assert(is_const(PEEK(1)));
+            // Peephole: inline constants.
+            PyObject *val = get_const(PEEK(1));
+            new_inst.opcode = _Py_IsImmortal(val) ? _LOAD_CONST_INLINE_BORROW : _LOAD_CONST_INLINE;
+            if (new_inst.opcode == _LOAD_CONST_INLINE) {
+                Py_INCREF(val);
+            }
+            new_inst.operand = (uintptr_t)val;
+            break;
+        }
+        case STORE_FAST_MAYBE_NULL:
+        case STORE_FAST: {
+            _Py_UOpsSymType *value = PEEK(1);
+            GETLOCAL(oparg) = value;
+            STACK_SHRINK(1);
+            break;
+        }
+        case COPY: {
+            _Py_UOpsSymType *bottom = PEEK(1 + (oparg - 1));
+            STACK_GROW(1);
+            PEEK(1) = bottom;
+            break;
+        }
+
+        case PUSH_NULL: {
+            STACK_GROW(1);
+            _Py_UOpsSymType *null_sym =  sym_init_push_null(ctx);
+            if (null_sym == NULL) {
+                goto error;
+            }
+            PEEK(1) = null_sym;
+            break;
+        }
+
+        case _INIT_CALL_PY_EXACT_ARGS: {
+            // Don't put in the new frame. Leave it be so that _PUSH_FRAME
+            // can extract callable, self_or_null and args later.
+            // This also means our stack pointer diverges from the real VM.
+            break;
+        }
+
+        case _PUSH_FRAME: {
+            int argcount = oparg;
+            // _INIT_CALL_PY_EXACT_ARGS's real stack effect in the VM.
+            stack_pointer += -1 - oparg;
+            // TOS is the new callable, above it self_or_null and args
+
+            PyFunctionObject *func = extract_func_from_sym(PEEK(1));
+            if (func == NULL) {
+                goto error;
+            }
+            PyCodeObject *co = (PyCodeObject *)func->func_code;
```

Comment:
I tried and wrote a comment above but it got lost. Here's what I wrote:

This isn't safe for my purposes. By the time the optimizer gets hold of the code objects, the code objects may be freed, due to triggering GC or whatever by Py_DECREF or object allocations. So it may hold an invalid reference.

This only works for Mark's peepholer because the peepholer does not allocate objects or decref things.

Reply:
Oh, I never saw that explanation before. It makes sense, but I'd still like to push back a little.

If we made both Abstract Context/Interpreter and Abstract Frame non-PyObjects (which looks pretty doable and will probably make the code less complex), there would be no other object allocations (only `PyMem_New()`, which should be safe). However, there will still be references to constants. We can have these owned by the the symbols in the arena, and we can DECREF them at the very end, when we're totally done.

I believe that if we do that, we are guaranteed no GC runs so all the code objects involved will be secure.

Now, once free threading and Tier 2 can coexist, we'll have a problem (another thread might free a code object), but my feeling is that we won't have to worry about that for 3.13 -- I think there are already other reasons why free threading will have to disable Tier 2 (and probably even some of the Tier 1 specializations). So I'm inclinde not to worry about this yet.

Reply:
I did that and it stall crashes. The problem is that constant propagation runs code, and a simple thing like `PyLong_Add` may trigger the GC. So this isn't feasible.